content/blog/dont-use-preludes-and-globs/index.md

```diff
@@ -0,0 +1,152 @@
+
+```rust
+//! # Examples
+//!
+//! ```
+//! # use mycrate::SomeType;
+//! # fn main() {
+//! let t = SomeType::new(); 
+//! # }
+//! ```
+```
+
+This will render as:
+
+```rust
+let t = SomeType::new();
+```
+
+### Reducing boilerplate
+
+Another common argument for preludes is that they reduce friction when importing types.
+
+I don't think that's true anymore. IDE support for Rust has made great strides in the last years. Nowadays, your IDE will just auto-import the type for you. (Largely thanks to [Rust Analyzer](https://rust-analyzer.github.io/) and [Rust Rover](https://www.jetbrains.com/rust/))
+Problem solved!
+
+And to also address the root cause: If you see the urge to add a prelude, because your crate requires importing many modules to be usable, maybe your public API is too big? Try to reduce the number modules.
+
+### Trait-only preludes
+
+Preludes, which only bring traits into scope might be acceptable.
+This trick is sometimes used in combination with extension traits, which add extra functionality to common types. `rayon` does that, for example, and it's quite magical if you can suddenly parallelize an iterator by just swapping `iter` with `par_iter`.
+
+From their documentation:
+
+```rust
+use rayon::prelude::*;
+fn sum_of_squares(input: &[i32]) -> i32 {
+    input.par_iter() // <-- just change that!
+         .map(|&i| i * i)
+         .sum()
+}
+```
+
+I think that's a solid use case for a prelude, but I never had such a clear-cut case in my own code.
+
+### Glob imports in tests
+
+The only exception for using glob imports I can think of is in tests. It's common to see `use super::*;` in tests to bring all functions from the parent module into scope. This can be a helpful mechanism to reduce boilerplate.
+
+```rust
+fn foo() -> i32 {
+    42
+}
+
+mod tests {
+    use super::*;
+    
+    #[test]
+    fn test_something() {
+        assert_eq!(foo(), 42);
+    }
+}
+```
+
+## Conclusion
+
+In your own code, avoid preludes and, by extension, glob imports.
+
+Here are the main disadvantages:
+
+- Makes it harder to know where types and functions come from
+- Can lead to naming conflicts, especially in larger codebases and when using multiple crates.
+- Complicates security audits because it's hard to see where code came from.
+- May hide module hierarchy and structure, which is useful if you want to learn how a crate is designed.
+- Can make IDE name resolution less reliable: if you have a name conflict, the IDE might not know which one you mean or might not be able to find the definition. 
+- Hides imports in documentation examples, making it harder to understand
+
+As you can see, the list of downsides is long. The only upside is that it saves a few keystrokes when writing code, which isn't worth it in the long run.
+
+**Pro tips:**
+
+- Enable the [clippy lint for wildcard imports](https://rust-lang.github.io/rust-clippy/master/index.html#/wildcard_imports) to catch glob imports in your code.
```

Comment:
New post idea: "My default clippy.toml for idiomatic rust"

Reply:
Heh, maybe not a full post, but an entry in a checklist.
Best thing I got so far is https://corrode.notion.site/Bootstrapping-Rust-Projects-f31a2fb063784a04bf334b69ffd8db99